### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by ericelliott** [ericelliott/isomorphic-express-boilerplate#4]
I'm planning to put a lot of love into this and make it an essential resource for anybody interested in building production Node and Isomorphic JavaScript applications. It's already very useful as-is, but we've been rewriting a lot of these things over and over again for every app, or searching through the haystack that is `npm` and spending too much time on tech selection instead of knowing what the best, most well-tested solutions are and running with them.

This boilerplate will eventually demonstrate a very simple production ready application, (basically "Hello, world!" including:

* [ ] User sign-in flows with multi-factor auth and robust password hashing using [Credential](https://github.com/ericelliott/credential)
* [ ] User authorization for requested functionality
* [ ] i18n
* [ ] Logging
* [ ] Error handling
* [ ] Routing
* [ ] Command manager (routes will trigger commands rather than kicking off rendering directly)
* [ ] Data modeling
* [ ] Data model validation
* [ ] HTML5 form validation using [h5v2](https://github.com/ericelliott/h5v2)
* [ ] CORS
* [ ] XSS prevention
* [ ] Security audit built-into automatic build validation
* [ ] Robust feature toggling with multivariate testing support
* [ ] Automatic hypermedia API route creation with [siren-resource](https://github.com/ericelliott/siren-resource)
* [ ] Full-featured build system

In other words, everything you'd want to reuse in most production-ready applications -- all done for you using years worth of tried and true best practices.

**Comment by zebulonj** [comment]
@ericelliott I'd like to help. I'm currently in the midst of a similar exercise as I work to break a legacy monolith down into micro-services... and as my first attempt at an isomorphic application I'm making up a lot as I go (so I don't pretend to have a lot of clear answers to offer). At the moment, the bit I'm most interested in is the isomorphic templates and rendering.

**Comment by ericelliott** [comment]
I have purposely not baked in a template system, but I'm considering changing that, using Minimongo, Track, and Blaze to do rendering on both the server and the client. It's a bit of o moving target, though, because the Meteor project doesn't seem to have any interest in maintaining an npm version of those libs.

Alternatively, we could show an example using React and Flux. Maybe those example implementations could be in a different repo which we link to from this one?

I'd love to hear your thoughts.

**Comment by capaj** [comment]
@ericelliott do you really think browserify is the way to go get trully isomorphic JS? To me it seems [JSPM](http://jspm.io/) is more up to the task, because it allows for any kind of modules, not just CommonJS. Much more easy to develop with, because there is no need for build steps.

**Comment by zebulonj** [comment]
Regarding rendering... I'd lean toward React and Flux.




—

**Comment by alexbasalyga** [comment]
@capaj I don't think the goal here is to use a technology that supports every kind of module, but simply to use the best technology available (in terms of simplicity of use and how well its maintained) to allow the use of npm as both the server and client package manager. JSPM very well may be the best tool for the job if the job is packaging CommonJS and AMD modules beside each other for the browser, but I think the job at hand is smaller in scope.

**Comment by ericelliott** [comment]
@capaj We're intersted in sharing the same code with the same source files authored in a single module format. JSPM may be a great solution for using ES6 modules in the browser, but I think it's a better fit to use Node style modules for code that is targeted at both Node and the browser.

**Comment by ericelliott** [comment]
@zebulonj That combo definitely has the mindshare right now.  :+1:

Should we be looking at trying to use JSX? Mixing markup into my JavaScript rubs me the wrong way.

I thought it was a great idea when PHP hit the scene. We all know how that played out... It was a bad idea then, and I'm not sure that anything has changed.

Also, if you're mixing markup into the JS, it seems like that puts an artificial limit targeting the output only to the DOM. IMO, one of the most attractive things about react is that it can target more than just the DOM...

**Comment by zebulonj** [comment]
@ericelliott I'm opening a new issue for this conversation about rendering (#13).

**Comment by h02e56** [comment]
@ericelliott thanks for spread your knowledge.:)
I've learned just seeing initial code. I would like to help someway.
About project, i am more "frontend" but i like going deeper on server side. I have played a little with express, although nowadays i like use tiny modules instead. 
About put the scripts on place i use to put them on the same node_modules folder, and now i feel weird having 2 node_modules folder. for sure is is just a matter of habit.:) 
I could give a hand with i18n , or routing with page or client bundling(think is solved)

**Comment by ivanoats** [comment]
"Full-featured build system" - Grunt, Gulp, or stick to NPM scripts? What do you think of Gulp? I could help with that.

**Comment by ericelliott** [comment]
h02e56: RE 2 node_modules folders -- this solution is much better than scattering relative paths all over the codebase that make refactoring harder than it needs to be.

@ivanoats npm is a placeholder until we have a consensus for Cloverfield. PLEASE DO implement a gulp build in [the todotasks repository](https://github.com/ericelliott/todotasks). See [the discussion in the Cloverfield repo](https://github.com/ericelliott/cloverfield/issues/2).

**Comment by terakilobyte** [comment]
I'd love to help on this project, and am looking forward to your courses. As an aside, how do you feel about babeljs?

**Comment by ericelliott** [comment]
Babel and similar projects that let us use future language features are awesome. Bonus if it exports node-compatible modules so you can `require()` with Node and Browserify.

**Comment by ericelliott** [comment]
By the way, did you see the announcement? [The first course has landed](https://www.kickstarter.com/projects/ericelliott/learn-javascript/posts/1151653)! We're letting students in and adding content. We're off and running. =)

**Comment by ivanoats** [comment]
@ericelliott I'd be happy to help but to clarify your request to implement gulp - isn't it already here? https://github.com/ericelliott/todotasks/tree/master/implementations/gulp

**Comment by ericelliott** [comment]
Oops! I completely forgot. Do you want to look it over and raise any issues or suggestions in that repo?

For now we're sticking with npm here unless / until we come to some consensus in Cloverfield. If you have anything to add to the [Cloverfield taskrunner discussions](https://github.com/ericelliott/cloverfield/issues/2), please add them.

One thing I like about a gulp implementation is that we have a better potential to make all the tasks completely cross-platform. I know we have some [windows compatibility issues](https://github.com/ericelliott/isomorphic-express-boilerplate/issues/9) it might help resolve.

Then again, Broccoli might do the same thing for us, and it may have some advantages over Gulp... but the Broccoli vs Gulp vs `x` debate belongs in the Cloverfield discussion thread, so we can concentrate on isomorphic / Node / Express issues here. =)

Do you want to create a gulp implementation branch here that reproduces all the functionality in a way that is more Windows friendly? I might be willing to merge it if it clearly works better than our current implementation.

**Comment by h02e56** [comment]
@ericelliott ```node_modules``` i was meaning putting them inside root node_modules
```node_modules/app/server```
```node_modules/app/client```

**Comment by ericelliott** [comment]
@h02e56 We don't put them into root `node_modules` because if you commit those to your repository, any time dependencies change, you get a ton of vendor noise in your pull requests, and code review becomes a bit of a nightmare.

I personally prefer that you make it easy to `rm -rf node_modules` if anything goes wrong with dependency updates. If you mix your modules with vendor modules, make changes in your modules, forget to commit, and then update dependencies, if something breaks, it can be hard to figure out what the working state was.

If your modules are not mixed with vendor modules and something goes wrong with a dependency update, you can safely `rm -rf node_modules && npm install` to get back to your working state -- even if you've made a bunch of changes and haven't committed them, yet.

That said, you should still commit often. ;)

**Comment by ericelliott** [comment]
Also, putting them in `app/node_modules` means that instead of typing `require('app/lib/foo');` you can just type `require('lib/foo');`

Just a few keystrokes per time, but it adds up.

**Comment by h02e56** [comment]
:+1:
"That said, you should still commit often." thats a true "como la copa de un pino" as we say here

**Comment by SOSANA** [comment]
are we using a flux pattern with react.js?

**Comment by nkbt** [comment]
Until there is a Relay, Flux is the way to go if you as me.
I've just rebuilt one of our angular modules during this sprint and first have done it without Flux at all. The code was boilerplaty as hell. Then added Flux and cleaned up to 30% of that mess right away. So I can't see any reasons not to use it as a default way to tackle data/app-state. No matter if it is React or angular.

Important question is if this particular boilerplate targets higher scope then just low-level-no-shared-state reusable React component or wants to handle more complex components? @ericelliott what's your idea on this?

**Comment by SOSANA** [comment]
@ericelliott  @nkbt have you guys checked out https://github.com/optimizely/nuclear-js ? its a Flux architecture built with ImmutableJS data structures.

**Comment by nkbt** [comment]
Sure thing. Check Redux by Dan Abramov, please

**Comment by ericelliott** [comment]
@SOSANA @nkbt as per the "state management" item on the Roadmap, we're going with [cerebral](https://github.com/christianalfoni/cerebral) for container components. Stateless components for UI primitives (buttons, comboboxes, toggle switches, etc...)

**Comment by nkbt** [comment]
Thanks. Sounds good

**Comment by SOSANA** [comment]
Interesting, I'll check it out

**Comment by firstdoit** [comment]
I would also vote for NuclearJS, as it seems far more mature than Cerebral, also relies heavily on ImmutableJS and does everything with little-to-no new concepts. Very lean, indeed.

**Comment by ericelliott** [comment]
@firstdoit @SOSANA I'm looking for a solution that has a working time-travel debugger -- not necessarily in the same module, but as part of a cohesive system. Nuclear's `console.log()` based debugging is cool, but seems considerably less convenient than the [cerebral debugger](https://www.youtube.com/watch?v=Fo86aiBoomE).

Because it's based on Immutable.js, it's conceivable that you could add a debugger like that to Nuclear, but AFAIK, that hasn't happened yet.

[Redux](https://github.com/gaearon/redux) is another state management library worth watching. It has features and architecture similar to Nuclear, and the author has [hinted at a time travel debugger on Twitter](https://twitter.com/_ericelliott/status/616018306454007808).

I'm on pins and needles...

**Comment by gaearon** [comment]
Give it a few days..

**Comment by ericelliott** [comment]
=)

**Comment by nkbt** [comment]
Redux is now in alpha. Probably the most FP-kind of flux. Give it a look
https://github.com/gaearon/redux/releases/tag/v1.0.0-alpha @gaearon is
doing very impressive job on that one.

**Comment by ericelliott** [comment]
:+1: yep. I'll be watching it closely. :)

**Comment by firstdoit** [comment]
@gaearon and @ericelliott, I appreciate the work on Redux greatly: a hot loadable flux implementation is pretty terrific. However, I feel it imposes lots of new concepts which the README does not explain that clearly. On the other hand, NuclearJS stores are *really* simple, with only two methods (`getInitialState` and `initialize`), and the only introduced concept is Getters, which are also fairly simple. @gaearon do you think redux could gravitate towards an API more similar to Nuclear? Or am I just missing the point here?

**Comment by gaearon** [comment]
>However, I feel it imposes lots of new concepts which the README does not explain that clearly

The proper docs will be a focus in the next couple of weeks. The current README wasn't intended to be this way; it's just the project got way more popular faster than I expected.

There is also a work progress in updating the API and terminology to make more sense: https://github.com/gaearon/redux/pull/195. Unfortunately the docs aren't updated there either. I'll have more time to work on that after the conference tomorrow.

>On the other hand, NuclearJS stores are really simple, with only two methods (getInitialState and initialize),

That sounds pretty fun to me, consider that “stores” in Redux are *just functions*. How can an object with two particular methods be simpler than a pure function? Especially considering that NuclearJS is tied to ImmutableJS, but Redux lets you use plain objects.

>do you think redux could gravitate towards an API more similar to Nuclear? 

I way I see see it, the API of Redux is simpler than the NuclearJS API. That's precisely the reason I started writing it, as I wasn't satisfied with the verbosity, the opinionatedness and the need to wrap everything in NuclearJS. In Redux, there is no `createThisStuff`, `createOtherStuff`, etc. Most of the time all you write are pure functions.

The docs are the real problem right now. We'll solve that. But I'd really love to hear your opinion on why you see the NuclearJS API being simpler than Redux.

**Comment by zebulonj** [comment]
My two-cents, I've found NuclearJS to be more complex (and definitely more opinionated) than either Redux or Cerebral. 

More importantly... I'd like to see this conversation focus as much on how the available options help to solve hard problems related to complex application state, as on what the API looks like. One of the most valuable contributions I see in Cerebral—and I say this without advocating for its adoption—is that it helps to address hard problems like state syncing.

Too often, I have chosen a framework based on API and a list of features (immutability, functional style,...), then found myself in the woods trying to solve those hard problems that don't emerge until you get past the TodoMVC level of complexity.

**Comment by ericelliott** [comment]
It's much easier for me to wrap my head around building undo/redo/history scrub features on an API of pure functions than NuclearJS methods. I'll be really excited to see the upcoming Redux talk and reworked docs. I'm willing to work with cerebral in the meantime for existing projects.

**Comment by firstdoit** [comment]
What I meant by *simplicity* was actually being *easier to use*, as in "there's more functionality/opinion baked in". I understand, now, that redux means to be very very agnostic, so everything should happen on userland (action dispatching, for example, via `bindActionCreators`). That's interesting and valid, but I wouldn't call it "simpler" for a newbie :)

**Comment by terakilobyte** [comment]
Just to chime in, I've been using Este as a jumping point for a few
personal projects. It might be worth looking at to take some ideas from.
https://github.com/steida/este

**Comment by vasco3** [comment]
What about including gzip support for html, js and css files ?
I implemented that in express without middleware, very leanly. 
Using unix after the npm build
https://github.com/vasco3/Apollo/blob/master/package.json#L10

And adding a route with proper encoding 
https://github.com/vasco3/Apollo/blob/master/app/app.js#L65-L69

**Comment by nkbt** [comment]
Isn't it something should be done on level of nginx/apache? I am personally
quite certain it should not exist on app level.

**Comment by vasco3** [comment]
well its not on the app level.. its on express which is the server.

**Comment by nkbt** [comment]
Better not to use node as production server for various reasons. For dev -
ok, but gzip is not needed for dev (usually)

**Comment by ericelliott** [comment]
@vasco3 Express has officially supported gzip middleware:

```js
var compress = require('compression');

app.use(compress()); 
```

I think if we're going to include compression, that's how it should be done. It's easy to see @nkbt's perspective, too, but I don't see the harm in adding two lines to the sourcecode. I do see the harm in ignoring gzip altogether. If we don't at least mention it, **people will forget about it.** And that would not be good. =)

**Comment by ericelliott** [comment]
I don't feel strongly about it, so I'd accept a PR that adds those two lines, or I'd accept a PR that adds a mention in the README and links to detailed instructions for both nginx and haproxy.

**Comment by vasco3** [comment]
will the compression middleware be compressing on every request?
Or do we still need to compress on `npm run build`?

**Comment by therealklanni** [comment]
It makes more sense to me to do this at the nginx level. It's probably faster at it, too.

**Comment by vasco3** [comment]
@therealklanni do you mean using nginx as a proxy for Node express server?

**Comment by vasco3** [comment]
What about cache busting?

**Comment by nkbt** [comment]
@vasco3 yes. Proxy for node server.

**Comment by ericelliott** [comment]
Yep. Everywhere I've pushed Node into production, we've had nginx & or haproxy, or both set up to reverse-proxy to our Node instances -- even for my smallest personal website deployment, I used nginx as a reverse proxy layer so I could route to different services easily.

**Comment by mycall** [comment]
Is it possible to have multiple versions of the boilerplate?  From most simple to advanced.

**Comment by ericelliott** [comment]
It's not only possible, it's likely. =)

**Comment by vasco3** [comment]
What are we considering for CSS? 
I've used https://github.com/sass/node-sass successfully.

**Comment by vasco3** [comment]
What about code coverage with codeCov.io ? :smiley:

**Comment by nkbt** [comment]
@vasco3 I prefer `coveralls` and use it for all my projects. Works well, but requires some minimal extra manual setup (set ENV vars in CI)

**Comment by vasco3** [comment]
@nkbt does **Coveralls** use xml like CircleCI or json? I had that discrepancy with CircleCI + CodeCov eg. https://github.com/vasco3/Oracul/blob/master/circle.yml#L18-L19
btw great job with the **tap-xunit** + CircleCI implementation

**Comment by nkbt** [comment]
Thanks, @vasco3!
Yep it does. Not XML, but normal lcov.info

```
- cat ${CIRCLE_ARTIFACTS}/coverage/lcov.info | ./node_modules/.bin/coveralls || true
```

https://github.com/nkbt/react-component-template/blob/master/circle.yml

**Comment by vasco3** [comment]
@nkbt how do you like isparta over babel-istanbul?

**Comment by nkbt** [comment]
Used isparta before babel-istanbul was there, works well, so I feel no need to swap it to something else for now.

**Comment by gacosta89** [comment]
Hey guys, do you plan to include Redux in the project? If that so, I can help you including react-devtools and the config for react hot reloading and reducers hot reloading.

**Comment by mycall** [comment]
react-kickstart uses hot-module-replacement ([babel-plugin-react-transform](https://github.com/gaearon/babel-plugin-react-transform) + [react-transform-hmr](https://github.com/gaearon/react-transform-hmr))

**Comment by ericelliott** [comment]
Yes, I plan to use Redux + HMR.

**Comment by nkbt** [comment]
HMR is added by #67

**Comment by gacosta89** [comment]
I think for react HMR you need this babel plugin: [**react-transform-hmr**](https://github.com/gaearon/react-transform-hmr), with this config: [**.babelrc**](https://github.com/gaearon/react-transform-hmr/#react).

**Comment by nkbt** [comment]
@gacosta89 that is how it is done in #67 =)

**Comment by gacosta89** [comment]
:( sorry.

**Comment by nkbt** [comment]
@gacosta89 no worries :ok_hand:

**Comment by guidsen** [comment]
What about writing a test for the reducers that have been added recently? Would be a great feature to show the users a good example on how the reducer can be tested (simple functions). Might have some time for it this week.

**Comment by ericelliott** [comment]
> What about writing a test for the reducers that have been added recently? Would be a great feature to show the users a good example on how the reducer can be tested (simple functions). Might have some time for it this week.

Sure. Please use tape and keep them simple.

**Comment by gacosta89** [comment]
Hey guys! what do you thing of [i18next](http://i18next.com/) for translations? I used it along with [react-i18next](https://github.com/i18next/react-i18next) in production. It has a very nice defaults and overrides system for keys.

**Comment by nkbt** [comment]
We used it in our prod app and moved to node-gettext for several reasons
(one of those - inability to do plurals in-place, so we had to have en.js
file, which is essentially wrong, and overall awkward syntax for templates
and plurals). In general I would recommend to stick to a standard
implementation - gettext. If you want some friendlier gettext solution,
check 'jed'
